seqr/views/utils/terra_api_utils.py

```diff
@@ -0,0 +1,226 @@
+        return self._make_request('post', path, headers, root_url, **kwargs)
+
+    def put(self, path, headers=None, root_url=None, **kwargs):
+        """See the __get() method."""
+        return self._make_request('put', path, headers, root_url, **kwargs)
+
+    def delete(self, path, headers=None, root_url=None):
+        """See the __get() method."""
+        return self._make_request('delete', path, headers, root_url)
+
+
+def is_google_authenticated(user):
+    if not TERRA_API_ROOT_URL:
+        return False
+    return len(user.social_auth.filter(provider = 'google-oauth2')) > 0
+
+
+_service_account_session = ServiceAccountSession()
+
+
+def sa_get_workspace_acl(workspace_namespace, workspace_name):
+    """
+    Request FireCloud access control list for workspace with a service account (sa).
+
+    Args:
+        workspace_namespace (str): namespace (name of billing project) of the workspace
+        workspace_name (str): the name of the workspace
+    Returns:
+        {
+            "user1Email": {
+              "accessLevel": "string",
+              "pending": true,
+              "canShare": true,
+              "canCompute": true
+            },
+            "user2Email": {
+              "accessLevel": "string",
+              "pending": true,
+              "canShare": true,
+              "canCompute": true
+            },
+            "user3Email": {
+              "accessLevel": "string",
+              "pending": true,
+              "canShare": true,
+              "canCompute": true
+            }
+          }
+          :param workspace_name:
+          :param workspace_namespace:
+    """
+    uri = "api/workspaces/{0}/{1}/acl".format(workspace_namespace, workspace_name)
+    r = _service_account_session.get(uri)
+    return json.loads(r.text)['acl']
+
+
+class BearerAuth(requests.auth.AuthBase):
+
+    def __init__(self, token):
+        self.token = token
+
+    def __call__(self, r):
+        r.headers["Authorization"] = "Bearer " + self.token
+        return r
+
+
+def _get_social(user):
+    social = user.social_auth.get(provider = 'google-oauth2')
+    if (social.extra_data['auth_time'] + social.extra_data['expires'] - 10) <= int(
+            time.time()):  # token expired or expiring?
+        strategy = load_strategy()
+        try:
+            social.refresh_token(strategy)
+        except Exception as ee:
+            logger.info('Refresh token failed. {}'.format(str(ee)))
+            raise TerraAPIException('Refresh token failed. {}'.format(str(ee)))
+    return social
+
+
+def _anvil_call(method, user, path, headers=None, root_url=None, **kwargs):
+    if not TERRA_API_ROOT_URL:
+        raise TerraAPIException('AnVIL access is not enabled')
+
+    social = _get_social(user)
+
+    # Todo: remove following two lines after whitelisted
+    request_func = getattr(_service_account_session, method)
+    return request_func(path, headers, root_url, **kwargs)
+
+    # Todo: remove the comment out
```

Comment:
Agree. All the `get/list_anvil_` functions work with a normal user. The `sa_get_` functions are the AnVIL calls with a service account. The functions for normal users are currently using the service account to simulate the user accounts. The reason for this is to do testing.

Reply:
I added checking for HTTP response `401` (Unauthorized). If it is the case return a blank `dict` rather than raise an exception. Temporarily, we can't access the permissions from AnVIL. But it keeps most of the seqr functions work without being whitelisted on AnVIL. We can simply remove this checking after whitelisted.